### ISSUE TITLE
Interpretation of MPI_Accumulate atomicity from v3.0 standard

### ISSUE THREAD

**Issue by sam6258** [open-mpi/ompi#5348]
### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)
v3.0

## Details of the problem

I'm looking for some insight into interpretation of the standard regarding MPI_Accumulate operations. The standard states in Section 11.3 that:

>  The outcome of concurrent conflicting accesses to the same memory locations is undefined; if  a location is updated by a put or  accumulate operation, then the outcome of loads or other RMA operations is undefined until the updating operation has completed at the target. There is one exception to this; namely, the same location can be updated by several concurrent accumulate calls, the outcome being as if these updates occurred in some order.

This section leads us to Section 11.7.1 regarding atomicity.

> The  outcome  of  concurrent  accumulate  operations  to  the  same  location  with  the  same
predefined  datatype  is  as  if  the  accumulates  were  done  at  that  location  in  some  serial
order.

Does "serial order" refer to the entire buffer being updated atomically or does each element of the buffer being atomically updated satisfy this requirement? 

For example:

I have a buffer containing an array of unsigned int's residing on rank 0. Each element of this buffer is initialized to an arbitrary value.
 
Rank 1 initiates an MPI_SUM operation using an array with each element initialized to 1 and targeting the buffer on rank 0.
Rank 2 initiates an MPI_REPLACE operation using an array with each element initialized to 0 and targeting the buffer on rank 0.
 
Case 1: If Rank 1's request is processed first followed by rank 2's request, we expect each element of the final buffer on rank 0 to contain 0.
Case 2: If Rank 2's request is processed first followed by rank 1's request, we expect each element of the final buffer on rank 0 to contain 1.
Case 3: If Rank 1's and Rank'2 requests are interleaved (as, for example, by the hardware), we expect the elements of the final buffer on rank 0 to contain a mixture of 0 and 1.
 
Does the standard allow for a mixture of 0 and 1 values in the final buffer as long as they were updated atomically as in case 3 above or is one of the first two cases above guaranteed?

**Comment by sam6258** [comment]
@jsquyres Do you have any insight on this? A fellow co-worker mentioned you are very well-versed in the standard and would probably know. Appreciate any help you can give. Thanks!

**Comment by bwbarrett** [comment]
I believe the intent of the MPI-3 RMA committee (of which I was a member, many years ago) was to support Case 3, but I'll admit the text is hard to parse explicitly.  Supporting this argument is the following text from MPI-3.1 Section 11.7.1:

> Concurrent accumulate operations with different origin and target pairs are
not ordered. Thus, there is no guarantee that the entire call to an accumulate operation is
executed atomically. 

This seems to imply strongly your case 3.  However, the next part is worrying:

> The effect of this lack of atomicity is limited: The previous correctness
conditions imply that a location updated by a call to an accumulate operation cannot be
accessed by a load or an RMA call other than accumulate until the accumulate operation has
completed (at the target).  Different interleavings can lead to different results only to the
extent that computer arithmetics are not truly associative or commutative. The outcome
of accumulate operations with overlapping types of different sizes or target displacements
is undefined.

This can be read to still support Case 3 for everything but `MPI_REPLACE`.  There are two possibilities here.  First, we (the MPI-3 RMA committee) meant for case 3 and forgot about `MPI_REPLACE`.  This is highly possible, given that the solution to the different operation restriction in MPI-2 was found late in the MPI-3 process and we were all tired at that point.  Second, we really meant either Case 1 or 2 would happen always, and really messed up this whole paragraph.  Personally, I'd read it as an application should be prepared for case 3: a mixed interleaving.

**Comment by jsquyres** [comment]
(I trust @bwbarrett's interpretation here -- I try to stay far, far away from all the MPI one-sided stuff 😉 )

**Comment by sam6258** [comment]
Thank you @bwbarrett and @jsquyres  Appreciate the responses and the reasoning. This explains enough to get me started on what I need to work on! I'm going to close, but may pop back here if I think of anything that needs additional clarification.
